[Merge to master] Major code cleanup

NISH1001 opened the pull request:
This PR will have the version `0.1.0` version of the framework. (Might have to rename the package name later or follow standard NASA open-source methodology for the releases). It's mostly a major cleanup and re-design of the original codebase, without changing any internal logic at best.

# Major Improvements
- old directories like `code/` and `minio-example` are moved to `sandbox` to retain legacy cofigs
- standard python packaging is done for the main framework at `evalit/` (Discussions at #8 )
- A base class at `evalit._base.AbstractAutomation` is introduced from which all the other automation components are inherited. (See `rclone.RcloneAutomation`, `nifi.NifiAutomation`)
- `rclone.rclone_automation.RcloneAutomation` now uses a more flexible `evalit.misc.shell.ShellExecutor` component to execute external commands.

# Minor Changes
- `loguru` is added to replace some print statements
- `debug` flag is added to rclone and nifi automation to be used later while debugging
- `tests/` directory has some dummy scripts to run the framework. We might have to add `examples/` directory?

# TODO
- [ ] further optimization and revamp is left
- [ ] better log parsers (Discussions at #10 )
- [x] license is to be added
- [ ] segregate transfer and analysis processes

---

@smarru @xhagrg @code-geek 
Can you please look at this PR and see if I am missing something. I just did few cleanups and improvements. There are many rooms for optimizations.

NISH1001 committed:
Cleanup code

Move legacy directories `code/nifi` and `minio-example` to `sandbox/`.

Move `requirements.txt` and pre commit hook config from `evaluation/` to
parent base.

NISH1001 committed:
Remove png files

NISH1001 committed:
Add setup.py for better packaging

Now we're using a placeholder package name called `evalit`. TODO: change
it later.

NISH1001 committed:
Add barebone abstract automation class

See `evalit._base.AbstractAutomation`. For now it loads yaml and
validates the keys.

NISH1001 committed:
Improve skeleton at evalit._base

NISH1001 committed:
Remove default yaml config loader with dictionary-based config in the
main constructor

Move the yaml to `AbstractAutomation.from_yaml` classmethod.

NISH1001 committed:
Refactor rclone automation class

Use `_base.AbstractAutomation` abstract base class to derive
`rclone.RcloneAutomation`.

NISH1001 committed:
Refactor (basic) nifi automation class

See `nifi.NifiAutomation`. It's now derived from
`_base.AbstractAutomation`.

NISH1001 committed:
Use external shell executor for rclone automation component

Now we use `misc.shell.ShellExecutor` component to execute arbitary
shell commands.

```python
rcloen_auto = RcloneAutomation(..., shell_executor=...)
```

NISH1001 committed:
Bump up the version to 0.1.0

NISH1001 committed:
Remove from_yaml classmethod in the base auotmation class

Now the `AbstractAutomation` constructor can take in both the path as
well as a dict to store configuration for source/dest.

NISH1001 committed:
Remove code/ directory

Pkoiralap commented on evalit/mft/mft_automation.py:
```diff
@@ -48,13 +50,15 @@ def run_automation(self):
         dest_storage_id = "d602acab-595a-44ac-a9b3-60c5ceca80e7"
```
is this supposed to be hardcoded? would it make more sense to put this in the env variable

NISH1001 replied:
Will have to check with @smarru as I don't have much idea on this. This PR is a code cleanup. So there's definitely more room for optimization/improvements. Thanks for pointing out this one.

Pkoiralap commented on evalit/mft/mft_automation.py:
```diff
@@ -48,13 +50,15 @@ def run_automation(self):
         end_time_map = {}
 
-        while(1):
+        while 1:
```
👀

NISH1001 replied:
We need to improve this. It's a legacy code. There's room for improvement here to avoid an infinite loop.

NISH1001 committed:
Refactor MFTAutomation class to use the new system design

Now `mft.mft_automation.MFTAutomation` is derived from
`_base.AbstractAutomation`.

Also, an external `misc.shell.ShellExecutor` object is used to invoke
system calls.

NISH1001 committed:
Add apache 2.0 license

code-geek commented on evalit/controller_test.py:
```diff
@@ -0,0 +1,69 @@
+file_list = ["testfile"]  # File list in source bucket
+
+# nifi_installation = "/sproj/MFT/nifi-1.15.3"
+nifi_installation = "/home/nishan/software/nifi/nifi-1.15.3"
```
is there a way we can avoid hardcoding this path?

NISH1001 replied:
Ya. I am planning to add env variables for that.

NISH1001 replied:
For file_list, I am planning to use s3 boto3 to get the list from source. That will be better I think.

code-geek commented on evalit/controller_test.py:
```diff
@@ -0,0 +1,69 @@
+generate_grapgs("rclone", rclone_result)
+caclulate_throughput(file_sizes, rclone_result)
+
+###### Airavata MFT ###########
+# automation = MFTAutomation(config_file, file_list, mft_installation)
+# mft_result = automation.run_automation()
+# print("Rclone automation results", mft_result)
+
+# generate_grapgs("mft", mft_result)
+# caclulate_throughput(file_sizes, mft_result)
```
if we don't need this anymore, consider removing

code-geek commented on evalit/mft/mft_automation.py:
```diff
@@ -0,0 +1,126 @@
+
+        # TODO: Optimize this!
+        while 1:
+            # Note: Why?
+            time.sleep(5)
```
Let's evaluate if this is necessary

NISH1001 commented on evalit/mft/mft_automation.py:
```diff
@@ -0,0 +1,126 @@
+        assert source_storage_id and dest_storage_id, "Invalid storage ids!"
+
+        njobs = max(njobs, 0)
+        njobs = min(njobs, multiprocessing.cpu_count())
```
I think this logic is wrong. will correct.

code-geek commented on evalit/misc/shell.py:
```diff
@@ -0,0 +1,61 @@
+@dataclass
+class ExecutionDTO:
+    """
+    Holds Shel execution data
```
nit: Shell

code-geek commented on evalit/odata/odata_automation.py:
```diff
@@ -1,9 +1,8 @@
         # Implementation here
-        return [[0, 10], [2, 30], [0, 50]]
\ No newline at end of file
+        return [[0, 10], [2, 30], [0, 50]]
```
some comments or variable names on why these numbers are this way would help someone taking a quick look at the code

NISH1001 replied:
I have removed these dummies. Thanks for pointing this out.

NISH1001 committed:
Bugfix njob issue in MFTAutomation
